onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorAttention.cpp

```diff
@@ -49,397 +37,501 @@ class DmlOperatorAttention : public DmlOperator
     DmlOperatorAttention(const MLOperatorKernelCreationContext& kernelCreationContext)
```

Comment:
>DmlOperatorAttention [](http://example.com/codeflow?start=4&length=20)

This is all incredibly complicated for an EP kernel, whereas ~99% of them just fill in some struct fields and ferry the parameters onward. Do you have any sense of whether a more official ONNX version of this op (not the contrib op) would settle on a practical default?

(I'm just thinking of what in the world this means for a WebNN attention op, which doesn't have any of the scaffolding for all this 😐).

Reply:
I think the plan (at the moment) is to make `MultiHeadAttention` part of the ONNX spec, but not `Attention`. `MultiHeadAttention` is a lot more straightforward and involves the core part that's the most computationally extensive. Attention just calculates the weights beforehand and then feed it to MHA, so in theory it's not needed (the ORT graph could very well handle it).